lib/virtual-fs/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+        self.opener.open(path.as_ref(), &self.conf)
+    }
+}
+
+/// This trait relies on your file closing when it goes out of scope via `Drop`
+#[typetag::serde(tag = "type")]
+pub trait VirtualFile: fmt::Debug + Send + Write + Read + Seek + 'static + Upcastable {
+    /// the last time the file was accessed in nanoseconds as a UNIX timestamp
+    fn last_accessed(&self) -> u64;
+
+    /// the last time the file was modified in nanoseconds as a UNIX timestamp
+    fn last_modified(&self) -> u64;
+
+    /// the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn created_time(&self) -> u64;
+
+    /// set the last time the file was accessed in nanoseconds as a UNIX timestamp
+    fn set_last_accessed(&self, _last_accessed: u64) {
+        debug!("{:?} did nothing in VirtualFile::set_last_accessed due to using the default implementation", self);
+    }
+
+    /// set the last time the file was modified in nanoseconds as a UNIX timestamp
+    fn set_last_modified(&self, _last_modified: u64) {
+        debug!("{:?} did nothing in VirtualFile::set_last_modified due to using the default implementation", self);
+    }
+
+    /// set the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn set_created_time(&self, _created_time: u64) {
+        debug!(
+            "{:?} did nothing in VirtualFile::set_created_time to using the default implementation",
+            self
+        );
+    }
+
+    /// the size of the file in bytes
+    fn size(&self) -> u64;
+
+    /// Change the size of the file, if the `new_size` is greater than the current size
+    /// the extra bytes will be allocated and zeroed
+    fn set_len(&mut self, new_size: u64) -> Result<(), FsError>;
+
+    /// Request deletion of the file
+    fn unlink(&mut self) -> Result<(), FsError>;
+
+    /// Store file contents and metadata to disk
+    /// Default implementation returns `Ok(())`.  You should implement this method if you care
+    /// about flushing your cache to permanent storage
+    fn sync_to_disk(&self) -> Result<(), FsError> {
+        Ok(())
+    }
+
+    /// Moves the file to a new location
+    /// NOTE: the signature of this function will change before stabilization
+    // TODO: stablizie this in 0.7.0 or 0.8.0 by removing default impl
+    fn rename_file(&self, _new_name: &std::path::Path) -> Result<(), FsError> {
+        panic!("Default implementation for now as this method is unstable; this default implementation or this entire method may be removed in a future release.");
+    }
+
+    /// Returns the number of bytes available.  This function must not block
+    fn bytes_available(&self) -> Result<usize, FsError>;
+
+    /// Used for polling.  Default returns `None` because this method cannot be implemented for most types
+    /// Returns the underlying host fd
+    fn get_raw_fd(&self) -> Option<i32> {
+        None
+    }
```

Comment:
This method is not used publicly, so I believe we can remove it.

At the same time by removing it we also assure that we shield from any specific host requirement in the Virtual FS.

Reply:
Small correction. Is indeed using in `state.rs` for polling.
I think we can abstract the polling as well into the Virtual FS.

That way we don't need to expose any host requirements in the VirtualFS itself.